File: android/app/src/main/java/com/sandnes/familyapp/ui/wishlist/WishlistPdf.kt
/*
 * Renders a wishlist to a shareable A4 PDF for family members who don't use the app.
 * Owner-facing export: it deliberately shows each wish's name, price, link and image —
 * NEVER reservation/claim status — so the surprise is preserved for the person exporting
 * their own list. Mirrors iOS `WishlistPDF`.
 *
 * The content-shaping logic ([wishlistPdfLines] / [wishlistPdfBlocks] / [paginateBlockHeights] /
 * [formatWishPrice] / [shortenedLink] / [sanitizedPdfFileName]) is pure and unit-tested (assert
 * the content lines, not pixels). The actual `android.graphics.pdf.PdfDocument` rendering in
 * [WishlistPdf.write] is the untested, pixel-producing shell around it. Images are fetched
 * best-effort with a per-image timeout — a failed image never fails the export.
 */
package com.sandnes.familyapp.ui.wishlist

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapShader
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Shader
import android.graphics.Typeface
import android.graphics.pdf.PdfDocument
import android.text.StaticLayout
import android.text.TextPaint
import coil3.SingletonImageLoader
import coil3.request.ImageRequest
import coil3.request.SuccessResult
import coil3.request.allowHardware
import coil3.toBitmap
import com.sandnes.familyapp.data.WishModel
import kotlinx.coroutines.withTimeoutOrNull
import java.io.File

/** The kind of line in the exported PDF — drives its font, colour and indent. */
enum class WishPdfLineKind { TITLE, SUBTITLE, BULLET, META }

/** One rendered line of the PDF. */
data class WishPdfLine(
    val text: String,
    val kind: WishPdfLineKind,
)

/** A group of lines kept together on one page (the header, or a single wish), plus its optional image. */
data class WishPdfBlock(
    val lines: List<WishPdfLine>,
    val imageUrl: String? = null,
)

/** Formats a wish's free-text price for the export: bare numbers become "299 kr" (NOK);
 *  anything else renders exactly as the user typed it. Pure — unit-tested. */
fun formatWishPrice(raw: String): String {
    val trimmed = raw.trim()
    val digitsOnly = trimmed.replace(" ", "")
    return if (digitsOnly.matches(Regex("^\\d+([.,]\\d{1,2})?$"))) "$trimmed kr" else trimmed
}

/** Compact display form of a wish link: scheme + www stripped, ellipsized. Pure — unit-tested. */
fun shortenedLink(
    url: String,
    maxLength: Int = 60,
): String {
    val stripped =
        url
            .trim()
            .removePrefix("https://")
            .removePrefix("http://")
            .removePrefix("www.")
            .removeSuffix("/")
    return if (stripped.length <= maxLength) stripped else stripped.take(maxLength - 1) + "…"
}

/**
 * The ordered content blocks of the export: block 0 is the header (title + optional subtitle),
 * then one block per non-blank wish (name + optional price/link lines + optional image).
 * Reservation status is never included. Pure — unit-tested.
 */
fun wishlistPdfBlocks(
    name: String,
    subtitle: String,
    wishes: List<WishModel>,
): List<WishPdfBlock> {
    val header = mutableListOf(WishPdfLine(name, WishPdfLineKind.TITLE))
    if (subtitle.isNotBlank()) header += WishPdfLine(subtitle, WishPdfLineKind.SUBTITLE)

    val wishBlocks =
        wishes
            .filter { it.text.isNotBlank() }
            .map { wish ->
                val lines = mutableListOf(WishPdfLine(wish.text, WishPdfLineKind.BULLET))
                wish.price?.takeIf { it.isNotBlank() }?.let { lines += WishPdfLine(formatWishPrice(it), WishPdfLineKind.META) }
                wish.link?.takeIf { it.isNotBlank() }?.let { lines += WishPdfLine(shortenedLink(it), WishPdfLineKind.META) }
                WishPdfBlock(lines, imageUrl = wish.imageUrl?.takeIf { it.isNotBlank() })
            }

    return listOf(WishPdfBlock(header)) + wishBlocks
}

/** Flattened content lines of the export (title, subtitle, wishes). Pure — unit-tested. */
fun wishlistPdfLines(
    name: String,
    subtitle: String,
    wishes: List<WishModel>,
): List<WishPdfLine> = wishlistPdfBlocks(name, subtitle, wishes).flatMap { it.lines }

/**
 * Greedy pagination: packs [heights] (one per block) onto pages of [contentHeight], never
 * splitting a block. Returns each page as the list of block indices it holds. Pure — unit-tested.
 */
fun paginateBlockHeights(
    heights: List<Float>,
    contentHeight: Float,
): List<List<Int>> {
    val pages = mutableListOf<MutableList<Int>>()
    var current = mutableListOf<Int>()
    var cursor = 0f
    heights.forEachIndexed { index, height ->
        if (current.isNotEmpty() && cursor + height > contentHeight) {
            pages += current
            current = mutableListOf()
            cursor = 0f
        }
        current += index
        cursor += height
    }
    if (current.isNotEmpty()) pages += current
    return pages
}

/** Filesystem-safe PDF file name from a wishlist name. Mirrors iOS `fileName(for:)`. Pure. */
fun sanitizedPdfFileName(name: String): String {
    val base = name.trim()
    val safe = base.ifEmpty { "Wishlist" }
    val separators = "/\\:?%*|\"<>"
    val cleaned = safe.map { if (it in separators) '-' else it }.joinToString("")
    return "$cleaned.pdf"
}

/** Renders and writes wishlist PDFs. The pure helpers above do the content shaping. */
object WishlistPdf {
    // A4 at 72 dpi, in points, matching iOS.
    private const val PAGE_WIDTH = 595
    private const val PAGE_HEIGHT = 842
    private const val MARGIN = 48
    private const val CONTENT_HEIGHT = (PAGE_HEIGHT - MARGIN * 2).toFloat()

    private const val TITLE_TEXT_SIZE = 26f
    private const val SUBTITLE_TEXT_SIZE = 13f
    private const val BULLET_TEXT_SIZE = 15f
    private const val META_TEXT_SIZE = 12f

    private const val META_INDENT = 0f
    private const val LINE_GAP = 3f // between lines within a block
    private const val BLOCK_GAP = 16f // after each wish block

    // Wish image thumbnail (points) and its gap to the text column.
    private const val IMAGE_SIZE = 90f
    private const val IMAGE_TEXT_GAP = 14f
    private const val IMAGE_CORNER_RADIUS = 10f
    private const val IMAGE_FETCH_PX = 360
    private const val IMAGE_FETCH_TIMEOUT_MS = 10_000L

    // Header accent rule under the title block (Heirloom Indigo).
    private const val ACCENT_COLOR = 0xFF5457E8.toInt()
    private const val ACCENT_HEIGHT = 3f
    private const val ACCENT_WIDTH = 64f
    private const val ACCENT_GAP = 10f

    // Fixed inks — the page is always white, so we must not use theme colours.
    private const val INK_COLOR = 0xFF000000.toInt()
    private const val META_COLOR = 0xFF525252.toInt() // grey ~ white 0.32

    private const val CACHE_SUBDIR = "wishlist_pdfs"

    /**
     * Writes the wishlist to a PDF in the app cache and returns the file (null on failure).
     * Suspends to download wish images (best-effort, per-image timeout). Call off the main
     * thread; share via [android.content.Intent.ACTION_SEND] through the app's FileProvider.
     */
    suspend fun write(
        context: Context,
        name: String,
        subtitle: String,
        wishes: List<WishModel>,
    ): File? =
        runCatching {
            val blocks = wishlistPdfBlocks(name, subtitle, wishes)
            val images = fetchImages(context, blocks)

            val contentWidth = PAGE_WIDTH - MARGIN * 2
            val heights = blocks.mapIndexed { index, block -> blockHeight(block, index in images.keys, contentWidth) }
            val pages = paginateBlockHeights(heights, CONTENT_HEIGHT)

            val document = PdfDocument()
            pages.forEachIndexed { pageIndex, blockIndices ->
                val pageInfo = PdfDocument.PageInfo.Builder(PAGE_WIDTH, PAGE_HEIGHT, pageIndex + 1).create()
                val page = document.startPage(pageInfo)
                var cursorY = MARGIN.toFloat()
                blockIndices.forEach { blockIndex ->
                    cursorY = drawBlock(page.canvas, blocks[blockIndex], images[blockIndex], cursorY, contentWidth)
                    if (blockIndex == 0) cursorY = drawAccentRule(page.canvas, cursorY)
                    cursorY += BLOCK_GAP
                }
                document.finishPage(page)
            }

            val dir = File(context.cacheDir, CACHE_SUBDIR).apply { mkdirs() }
            val file = File(dir, sanitizedPdfFileName(name))
            file.outputStream().use { document.writeTo(it) }
            document.close()
            file
        }.getOrNull()

    /** Downloads each block's image (software bitmap — PDF canvases can't draw hardware ones).
     *  Failures and timeouts are silently skipped so the export always succeeds. */
    private suspend fun fetchImages(
        context: Context,
        blocks: List<WishPdfBlock>,
    ): Map<Int, Bitmap> {
        val loader = SingletonImageLoader.get(context)
        val result = mutableMapOf<Int, Bitmap>()
        blocks.forEachIndexed { index, block ->
            val url = block.imageUrl ?: return@forEachIndexed
            val bitmap =
                runCatching {
                    withTimeoutOrNull(IMAGE_FETCH_TIMEOUT_MS) {
                        val request =
                            ImageRequest
                                .Builder(context)
                                .data(url)
                                .size(IMAGE_FETCH_PX, IMAGE_FETCH_PX)
                                .allowHardware(false)
                                .build()
                        (loader.execute(request) as? SuccessResult)?.image?.toBitmap()
                    }
                }.getOrNull()
            if (bitmap != null) result[index] = bitmap
        }
        return result
    }

    /** Draws one block (optional thumbnail left, text right) starting at [top]; returns the Y just below it. */
    private fun drawBlock(
        canvas: Canvas,
        block: WishPdfBlock,
        image: Bitmap?,
        top: Float,
        contentWidth: Int,
    ): Float {
        val textIndent = if (image != null) IMAGE_SIZE + IMAGE_TEXT_GAP else 0f
        val textWidth = (contentWidth - textIndent).toInt().coerceAtLeast(1)

        if (image != null) {
            drawRoundedImage(canvas, image, MARGIN.toFloat(), top)
        }

        var cursorY = top
        block.lines.forEachIndexed { index, line ->
            if (index > 0) cursorY += LINE_GAP
            cursorY += drawLine(canvas, line, cursorY, textIndent, textWidth)
        }
        // The block occupies at least the thumbnail's height.
        return if (image != null) maxOf(cursorY, top + IMAGE_SIZE) else cursorY
    }

    /** Center-crops [bitmap] into a rounded IMAGE_SIZE square at ([left], [top]). */
    private fun drawRoundedImage(
        canvas: Canvas,
        bitmap: Bitmap,
        left: Float,
        top: Float,
    ) {
        val shader = BitmapShader(bitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
        val scale = IMAGE_SIZE / minOf(bitmap.width, bitmap.height).toFloat()
        val matrix =
            Matrix().apply {
                setScale(scale, scale)
                postTranslate(
                    left - (bitmap.width * scale - IMAGE_SIZE) / 2f,
                    top - (bitmap.height * scale - IMAGE_SIZE) / 2f,
                )
            }
        shader.setLocalMatrix(matrix)
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply { this.shader = shader }
        val rect = RectF(left, top, left + IMAGE_SIZE, top + IMAGE_SIZE)
        canvas.drawRoundRect(rect, IMAGE_CORNER_RADIUS, IMAGE_CORNER_RADIUS, paint)
    }

    /** Short Heirloom-Indigo rule under the header; returns the Y just below it. */
    private fun drawAccentRule(
        canvas: Canvas,
        top: Float,
    ): Float {
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = ACCENT_COLOR }
        val y = top + ACCENT_GAP
        canvas.drawRoundRect(
            RectF(MARGIN.toFloat(), y, MARGIN + ACCENT_WIDTH, y + ACCENT_HEIGHT),
            ACCENT_HEIGHT / 2f,
            ACCENT_HEIGHT / 2f,
            paint,
        )
        return y + ACCENT_HEIGHT
    }

    /** Draws one wrapped line at [top]; returns the height it consumed. */
    private fun drawLine(
        canvas: Canvas,
        line: WishPdfLine,
        top: Float,
        indent: Float,
        width: Int,
    ): Float {
        val layout = layoutFor(line, width)
        canvas.save()
        canvas.translate(MARGIN + indent + if (line.kind == WishPdfLineKind.META) META_INDENT else 0f, top)
        layout.draw(canvas)
        canvas.restore()
        return layout.height.toFloat()
    }

    /** Measures a block's total height (used for pagination). */
    private fun blockHeight(
        block: WishPdfBlock,
        hasImage: Boolean,
        contentWidth: Int,
    ): Float {
        val textIndent = if (hasImage) IMAGE_SIZE + IMAGE_TEXT_GAP else 0f
        val width = (contentWidth - textIndent).toInt().coerceAtLeast(1)
        var height = 0f
        block.lines.forEachIndexed { index, line ->
            if (index > 0) height += LINE_GAP
            height += layoutFor(line, width).height
        }
        val accent = if (block.lines.firstOrNull()?.kind == WishPdfLineKind.TITLE) ACCENT_GAP + ACCENT_HEIGHT else 0f
        return maxOf(height, if (hasImage) IMAGE_SIZE else 0f) + accent
    }

    private fun layoutFor(
        line: WishPdfLine,
        width: Int,
    ): StaticLayout {
        val paint = paintFor(line.kind)
        return StaticLayout.Builder.obtain(line.text, 0, line.text.length, paint, width).build()
    }

    private fun paintFor(kind: WishPdfLineKind): TextPaint =
        TextPaint(TextPaint.ANTI_ALIAS_FLAG).apply {
            when (kind) {
                WishPdfLineKind.TITLE -> {
                    textSize = TITLE_TEXT_SIZE
                    color = INK_COLOR
                    typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
                }
                WishPdfLineKind.SUBTITLE -> {
                    textSize = SUBTITLE_TEXT_SIZE
                    color = META_COLOR
                    typeface = Typeface.DEFAULT
                }
                WishPdfLineKind.BULLET -> {
                    textSize = BULLET_TEXT_SIZE
                    color = INK_COLOR
                    typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
                }
                WishPdfLineKind.META -> {
                    textSize = META_TEXT_SIZE
                    color = META_COLOR
                    typeface = Typeface.DEFAULT
                }
            }
        }
}

File: ios/FamilyApp/Features/Wishlist/WishlistPDF.swift
// Renders a wishlist to a shareable A4 PDF for family members who don't use the app.
// Owner-facing export: it deliberately shows each wish's name, price, link and image —
// never reservation/claim status — so the surprise is preserved for the person exporting
// their own list. Pure UIKit (UIGraphicsPDFRenderer), unit-tested via WishlistPDFTests.
// Mirrors Android `WishlistPdf`. Images are fetched best-effort with a per-image timeout;
// a failed image never fails the export.
import UIKit

/// Formats a wish's free-text price for the export: bare numbers become "299 kr" (NOK);
/// anything else renders exactly as the user typed it. Pure — unit-tested.
func formatWishPrice(_ raw: String) -> String {
    let trimmed = raw.trimmingCharacters(in: .whitespaces)
    let digitsOnly = trimmed.replacingOccurrences(of: " ", with: "")
    let isBareNumber = digitsOnly.range(of: #"^\d+([.,]\d{1,2})?$"#, options: .regularExpression) != nil
    return isBareNumber ? "\(trimmed) kr" : trimmed
}

/// Compact display form of a wish link: scheme + www stripped, ellipsized. Pure — unit-tested.
func shortenedLink(_ url: String, maxLength: Int = 60) -> String {
    var stripped = url.trimmingCharacters(in: .whitespaces)
    for prefix in ["https://", "http://"] where stripped.hasPrefix(prefix) {
        stripped.removeFirst(prefix.count)
    }
    if stripped.hasPrefix("www.") { stripped.removeFirst(4) }
    if stripped.hasSuffix("/") { stripped.removeLast() }
    guard stripped.count > maxLength else { return stripped }
    return String(stripped.prefix(maxLength - 1)) + "…"
}

enum WishlistPDF {
    private static let pageSize = CGSize(width: 595.2, height: 841.8) // A4 at 72 dpi
    private static let margin: CGFloat = 48
    // Fixed dark inks — the page is always white, so we must NOT use dynamic colors like
    // .label/.secondaryLabel: in the app's dark mode those resolve near-white and vanish.
    private static let inkColor = UIColor.black
    private static let metaColor = UIColor(white: 0.32, alpha: 1)
    // Heirloom Indigo accent rule under the header (fixed, not a dynamic theme color).
    private static let accentColor = UIColor(red: 0x54 / 255.0, green: 0x57 / 255.0, blue: 0xE8 / 255.0, alpha: 1)

    private static let imageSize: CGFloat = 90
    private static let imageTextGap: CGFloat = 14
    private static let imageFetchTimeout: TimeInterval = 10
    private static let blockGap: CGFloat = 16

    /// Downloads wish images (best-effort) and writes the wishlist to a temp-directory PDF.
    /// Returns its file URL, or nil on failure. `subtitle` (e.g. "By Alice") is passed in
    /// already localized so this stays free of the main-actor `L` helper.
    static func make(name: String, subtitle: String, wishes: [WishModel]) async -> URL? {
        let images = await fetchImages(for: wishes)
        return render(name: name, subtitle: subtitle, wishes: wishes, images: images)
    }

    /// Fetches each wish's image; failures and timeouts are skipped so the export always succeeds.
    private static func fetchImages(for wishes: [WishModel]) async -> [String: UIImage] {
        var result: [String: UIImage] = [:]
        for wish in wishes {
            guard let raw = wish.imageUrl?.trimmingCharacters(in: .whitespaces), !raw.isEmpty,
                  let url = URL(string: raw)
            else { continue }
            let request = URLRequest(url: url, timeoutInterval: imageFetchTimeout)
            if let (data, _) = try? await URLSession.shared.data(for: request),
               let image = UIImage(data: data) {
                result[wish.id] = image
            }
        }
        return result
    }

    /// Synchronous rendering shell around the fetched images.
    private static func render(
        name: String, subtitle: String, wishes: [WishModel], images: [String: UIImage]
    ) -> URL? {
        let pageRect = CGRect(origin: .zero, size: pageSize)
        let contentWidth = pageSize.width - margin * 2
        let renderer = UIGraphicsPDFRenderer(bounds: pageRect)
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent(fileName(for: name))

        do {
            try renderer.writePDF(to: url) { ctx in
                ctx.beginPage()
                var cursorY = margin

                cursorY += draw(
                    name, at: CGPoint(x: margin, y: cursorY), width: contentWidth,
                    font: .systemFont(ofSize: 26, weight: .bold), color: Self.inkColor
                )
                if !subtitle.isEmpty {
                    cursorY += 6
                    cursorY += draw(
                        subtitle, at: CGPoint(x: margin, y: cursorY), width: contentWidth,
                        font: .systemFont(ofSize: 13, weight: .regular), color: Self.metaColor
                    )
                }
                // Short indigo accent rule under the header.
                cursorY += 10
                let accent = UIBezierPath(
                    roundedRect: CGRect(x: margin, y: cursorY, width: 64, height: 3), cornerRadius: 1.5
                )
                accentColor.setFill()
                accent.fill()
                cursorY += 3 + 24

                let bodyFont = UIFont.systemFont(ofSize: 15, weight: .semibold)
                let metaFont = UIFont.systemFont(ofSize: 12, weight: .regular)
                for wish in wishes where !wish.text.trimmingCharacters(in: .whitespaces).isEmpty {
                    let image = images[wish.id]
                    let needed: CGFloat = image != nil ? imageSize + blockGap : 40
                    cursorY = startPageIfNeeded(cursorY, ctx: ctx, needed: needed)

                    let blockTop = cursorY
                    let textX = margin + (image != nil ? imageSize + imageTextGap : 0)
                    let textWidth = contentWidth - (image != nil ? imageSize + imageTextGap : 0)

                    if let image {
                        drawSquare(image, at: CGPoint(x: margin, y: blockTop))
                    }

                    cursorY += draw(
                        wish.text, at: CGPoint(x: textX, y: cursorY), width: textWidth,
                        font: bodyFont, color: Self.inkColor
                    )
                    if let price = wish.price, !price.trimmingCharacters(in: .whitespaces).isEmpty {
                        cursorY += 3
                        cursorY += draw(
                            formatWishPrice(price), at: CGPoint(x: textX, y: cursorY), width: textWidth,
                            font: metaFont, color: Self.metaColor
                        )
                    }
                    if let link = wish.link, !link.trimmingCharacters(in: .whitespaces).isEmpty {
                        cursorY += 3
                        cursorY += draw(
                            shortenedLink(link), at: CGPoint(x: textX, y: cursorY), width: textWidth,
                            font: metaFont, color: Self.metaColor
                        )
                    }
                    // The block occupies at least the thumbnail's height.
                    if image != nil { cursorY = max(cursorY, blockTop + imageSize) }
                    cursorY += blockGap
                }
            }
            return url
        } catch {
            return nil
        }
    }

    /// Center-crops [image] into an `imageSize` square at [origin].
    private static func drawSquare(_ image: UIImage, at origin: CGPoint) {
        let rect = CGRect(x: origin.x, y: origin.y, width: imageSize, height: imageSize)
        guard let ctx = UIGraphicsGetCurrentContext() else { return }
        ctx.saveGState()
        UIBezierPath(rect: rect).addClip()
        let scale = max(imageSize / image.size.width, imageSize / image.size.height)
        let drawSize = CGSize(width: image.size.width * scale, height: image.size.height * scale)
        let drawOrigin = CGPoint(
            x: rect.midX - drawSize.width / 2,
            y: rect.midY - drawSize.height / 2
        )
        image.draw(in: CGRect(origin: drawOrigin, size: drawSize))
        ctx.restoreGState()
    }

    /// Draws wrapped text and returns the height it consumed.
    private static func draw(
        _ text: String, at point: CGPoint, width: CGFloat, font: UIFont, color: UIColor
    ) -> CGFloat {
        let paragraph = NSMutableParagraphStyle()
        paragraph.lineBreakMode = .byWordWrapping
        let attrs: [NSAttributedString.Key: Any] = [
            .font: font, .foregroundColor: color, .paragraphStyle: paragraph,
        ]
        let bounds = CGSize(width: width, height: .greatestFiniteMagnitude)
        let rect = (text as NSString).boundingRect(
            with: bounds, options: [.usesLineFragmentOrigin, .usesFontLeading],
            attributes: attrs, context: nil
        )
        (text as NSString).draw(
            with: CGRect(x: point.x, y: point.y, width: width, height: ceil(rect.height)),
            options: [.usesLineFragmentOrigin, .usesFontLeading], attributes: attrs, context: nil
        )
        return ceil(rect.height)
    }

    /// Starts a new page (resetting the cursor to the top margin) when the next block
    /// wouldn't fit above the bottom margin.
    private static func startPageIfNeeded(
        _ cursorY: CGFloat, ctx: UIGraphicsPDFRendererContext, needed: CGFloat
    ) -> CGFloat {
        if cursorY + needed > pageSize.height - margin {
            ctx.beginPage()
            return margin
        }
        return cursorY
    }

    private static func fileName(for name: String) -> String {
        let base = name.trimmingCharacters(in: .whitespacesAndNewlines)
        let safe = base.isEmpty ? "Wishlist" : base
        let cleaned = safe.components(separatedBy: CharacterSet(charactersIn: "/\\:?%*|\"<>"))
            .joined(separator: "-")
        return "\(cleaned).pdf"
    }
}
